# Flex Meta-Storms

![Version](https://img.shields.io/badge/Version-1.0%20-brightgreen)
![Release date](https://img.shields.io/badge/Released%20date-Sept.%2021%2C%202022-brightgreen)



# Contents

- [Introduction](#introduction)
- [System Requirement and dependency](#system-requirement-and-dependency)
- [Installation guide](#installation-guide)
- [Usage](#usage)
- [Example dataset](#example-dataset)
- [Tools in this package](#tools-in-this-package)
- [Supplementary](#supplementary)

  

# Introduction

Flex Meta-Storms (FMS) algorithm implements the “local alignment” concept of microbiome beta diversity for the first time. With microbes of interest (e.g. biomarkers), FMS produces a normalized phylogenetic distance between microbiome pairs. Therefore, FMS can mine potential relationships generated by a small subset of microbes in community samples, elucidating beta diversity with greater sensitivity and flexibility, thereby contributing to a deeper understanding of microbe-host interactions effect. FMS is developed as a standalone package, and it is also available as a plug-in for Parallel-Meta Suite.

# System Requirement and dependency

## Hardware Requirements

Flex Meta-Storms only requires a standard computer with sufficient RAM to support the operations defined by a user. For typical users, this would be a computer with about 2 GB of RAM. For optimal performance, we recommend a computer with the following specs:

  RAM: 8+ GB  
  CPU: 4+ cores

## Software Requirements

OpenMP library is the C/C++ parallel computing library. Most Linux releases have OpenMP already been installed in the system. In Mac OS X, to install the compiler that supports OpenMP, we recommend using the Homebrew package manager:
```
brew install gcc
```

## Rscript environment (optional for biomarker selection)

For biomarkers selection (optional), FMS requires cran-R (http://cran.r-project.org/) 3.2 or higher for the execution of “.R” scripts. Then all packages could be automatically installed and updated by the FMS installer.

# Installation guide

## Automatic Installation (recommended)

At present, Flex Meta-Storms provides a fully automatic installer for easy installation.

**a. Download the package**

```
git clone https://github.com/qdu-bioinfo/flex-meta-storms.git	
```
Click here to download the Parallel-Meta Suite plug-in version

**b. Install by installer**
```
cd flex-meta-storms
source install.sh
```

The package should take less than 5 minutes to install on a computer with the specifications recommended above.

The example dataset could be found at “example” folder. Check the “example/Readme” for details about the demo run.

## Manual Installation

If the automatic installer fails, Flex Meta-Storms can still be installed manually.

**a. Download the package**
```
git clone https://github.com/qdu-bioinfo/flex-meta-storms.git	
```

**b. Configure the environment variables (the default environment variable configuration file is “~/.bashrc”)**
```
export FlexMetaStorms=Path to Flex Meta-Storms
export PATH=”$PATH:$FlexMetaStorms/bin/”
export PATH=”$PATH:$FlexMetaStorms/Rscript/”
source ~/.bashrc
```
**c. Compile the source code**

```
cd flex-meta-storms
make
```
**d. Install R packages (optional for biomarker selection)**

```
Rscript $ FlexMetaStorms/Rscript/config.R
```

# Usage
**a.  Calculate Flex Meta-Storms distance**
- User installed as a software
```
FMS-comp-taxa -T dataset.abd -M 0 -m bio_marker.tab -o target.dist
```
The output file “target.dist” is the pairwise distance matrix. 
The format of "bio_marker. tab" is as follows:
|  | A.test | B.test | ... |
| :----:| :----: | :----: | :----: |
| OTU_1| ... | ... | ... |
| OTU_2 | ... | ... | ... |
| OTU_3 | ... | ... | ... |
| OTU_4 | ... | ... | ... |
| ... | ... | ... | ... | ... |

- User installed as a PMS plugin
```
PM-comp-taxa-local -T dataset.abd -M 0 -m bio_marker.tab -o target.dist
```
The output file “target.dist” is the pairwise distance matrix. 

**b. Calculate exact markers distance**
- User installed as a software
```
FMS-comp-taxa -T dataset.abd -M 1 -m bio_marker.tab -o exact.dist
```
The output file “exact.dist” is the pairwise distance matrix. 

- User installed as a PMS plugin
```
PM-comp-taxa-local -T dataset.abd -M 1 -m bio_marker.tab -o exact.dist
```
The output file “exact.dist” is the pairwise distance matrix. 

# Example dataset
Here, we provide a demo dataset (Real dataset I) in the "examples" folder with real species information for 136 individuals. In this package, "dataset.meta" is the meta information of the samples, and "dataset.abd" is the relative abundance at the OTU level.
To run the demo, you can either:
```
cd example
sh Readme
```
or type the following command to calculate the Flex Meta-Storms distance:
- User installed as a software
```
PM_Marker_Test.R -m dataset.meta -i dataset.abd -o Marker

FMS-comp-taxa -T dataset.abd -M 0 -m ./Marker/Out.Group.sig.meanTests.xls -o target.dist
```
The output file “target.dist” is the pairwise distance matrix. 

- User installed as a PMS plugin
```
PM_Marker_Test.R -m dataset.meta -i dataset.abd -o Marker

PM-comp-taxa-local -T dataset.abd -M 0 ./Marker/Out.Group.sig.meanTests.xls -o target.dist
```

This demo run should take less than 5 minutes on a recommended computer.

# Tools in this package
**a. PM_Marker_Test.R**

Screening biomarkers for community samples. Run:
```
PM_Marker_Test.R -h
```
for detailed parameters.


**b. FMS-comp-taxa or PM-comp-taxa-local**

Calculate Exact markers distance or Flex Meta-Storms distance between samples. Run:
```
FMS-comp-taxa -h 
or
PM-comp-taxa-local -h
```
for detailed parameters.



# Supplementary

[Real Dataset 1](http://) Contains 88 autism samples inferred from 16S rRNA genes by Parallel Meta-Suite.

[Real Dataset 2](http://) Contains 104 colorectal cancer samples inferred from 16S rRNA genes by Parallel Meta-Suite.




